Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm

```diff
@@ -127,6 +128,36 @@
     return result;
 }
 
+bool WebExtensionContextProxy::hasPermission(const String& permission) const
+{
+    WallTime currentTime = WallTime::now();
+
+    // If the next expiration date hasn't passed yet, there is nothing to remove.
+    if (m_nextGrantedPermissionsExpirationDate != WallTime::nan() && m_nextGrantedPermissionsExpirationDate > currentTime)
+        goto finish;
+
+    m_nextGrantedPermissionsExpirationDate = WallTime::infinity();
+
+    m_grantedPermissions.removeIf([&](auto& entry) {
+        if (entry.value <= currentTime)
+            return true;
+
+        if (entry.value < m_nextGrantedPermissionsExpirationDate)
+            m_nextGrantedPermissionsExpirationDate = entry.value;
+
+        return false;
+    });
+
+finish:
+    return m_grantedPermissions.contains(permission);
+}
+
+void WebExtensionContextProxy::updateGrantedPermissions(PermissionsMap&& permissions)
+{
+    m_grantedPermissions = WTFMove(permissions);
+    m_nextGrantedPermissionsExpirationDate = WallTime::nan();
```

Comment:
Will `m_nextGrantedPermissionsExpirationDate` be initialized to `nan` when the object is created? Should we do that?

Reply:
Yes it is in the header.